stage #0/module #4 Algorithms, Data Structures/algorithms_and_data_structures.md

```diff
@@ -0,0 +1,487 @@
+```
+int[] arr = {1, 10, 100};
+int a = arr.length;
+System.out.println(a);
+// Output: 3
+```
+
+#### Accessing Array's Elements
+In order to access an element of an array, we should write an array's variable and an element's index encapsulated in square braces.
+![Array Initialization](img/array_initialization_curly_braces.png)
+```
+int element = arr[5];
+System.out.println(element);
+// Output: 27
+```
+When accessing an element, if the index is negative or greater or equal to "length", Java will throw an "ArrayIndexOutOfBoundException".
+
+#### Assigning a value to array's elements
+We can assign a value to an array's element by writing an array's variable name followed by an element's index, then use an assignment operator "=" and define a compatible value.
+```
+int[] arr = new int[6];
+arr[5] = 27;
+```
+![Assigning a value](img/array_assigning_value.png)
+
+
+#### Iterating Through an Array
+We can iterate through an array with the help of an array's "length" property and "for", "while", "do-while" statements.
+```
+int[] ints = {3, 2, 1};
+
+// "for" statement
+for (int i = 0; i < ints.length; i++) {
+    int element = ints[i];
+    System.out.print(element);
+}
+
+// "while" statement
+int index = 0;
+while (index < ints.length) {
+    int element = ints[i];
+    System.out.print(element);
+    index++;
+}
+
+// "do-while" statement
+int i = 0;
+do {
+    System.out.print(ints[i]);
+    i++;
+} while (index < ints.length);
+
+// Output:
+123123123
+```
+
+#### Copying an Array
+The System class has an arraycopy method that you can use to efficiently copy data from one array into another:
+```
+public static void arraycopy(Object src, int srcPos, Object dest, int destPos, int length)
+```
+The 1st Object argument specifies an array "to copy from", the second one defines an array "to copy to". An argument "srcPos" specifies the starting position in the array "to copy from" (source array), "destPos" defines the starting position in the array "to copy to" (destination array), "length" is the number of array's elements, which should be copied.
+```
+int[] src = {798, 33, 21, 79, 12, 987};
+int[] dest = new int[9];
+System.arraycopy(src, 0, dest, 2, 5)
+```
+![System.arraycopy(..)](img/arraycopy.png)
+
+#### Multidimensional Arrays
+Arrays we have mentioned till now are called one-dimensional arrays. However, we can declare multidimensional arrays in Java. A multidimensional array in Java is an array of arrays, that is, each element of a multidimensional array is an array itself.
+
+A two-dimensional array is the simplest form of multi-dimensional arrays, in other words, it is an array of one-dimensional arrays. it is stored in the form of rows and columns.
```

Comment:
feels like the lack of an example (for me it would be simpler to understand that if I could see at least a 2-dimentional coordinate plane for 2D array)